C++ APIs TransformerEncoder

glaringlee opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#43187 C++ APIs TransformerEncoder**

Differential Revision: [D23182770](https://our.internmc.facebook.com/intern/diff/D23182770)

Commit:
C++ APIs TransformerEncoder

[ghstack-poisoned]

Commit:
Update on "[WIP] C++ APIs TransformerEncoder"


Differential Revision: [D23182770](https://our.internmc.facebook.com/intern/diff/D23182770)

[ghstack-poisoned]

Commit:
Update on "[WIP] C++ APIs TransformerEncoder"


Differential Revision: [D23182770](https://our.internmc.facebook.com/intern/diff/D23182770)

[ghstack-poisoned]

Commit:
Update on "[WIP] C++ APIs TransformerEncoder"


Differential Revision: [D23182770](https://our.internmc.facebook.com/intern/diff/D23182770)

[ghstack-poisoned]

Commit:
Update on "C++ APIs TransformerEncoder"


Differential Revision: [D23182770](https://our.internmc.facebook.com/intern/diff/D23182770)

[ghstack-poisoned]

zou3519 commented on torch/csrc/api/include/torch/nn/modules.h:
```diff
@@ -31,3 +31,4 @@
 #include <torch/nn/modules/activation.h>
 #include <torch/nn/modules/normalization.h>
 #include <torch/nn/modules/transformerlayer.h>
+#include <torch/nn/modules/transformercoder.h>
```
nit: we should probably call it `transformerencoder.h`. I know the name is a bit longer but it improves discovery (and also the word "coder" can mean a lot of things)

zhangguanheng66 replied:
good catch. or similarly to the python one, can we put everything as `transformer.h`?

glaringlee replied:
Ah, let me explain. We can not put everything into transformer.h, it will cause `cycle include` otherwise. This is due to the implementation design of the python version, since python doesn't have such problem.

Let me use encoder as an example, decoder has the same thing.
In C++ APIs, a `Module` class' constructor takes an `Options` class at the input and the `Options` class contains all the initialization parameters. TransformerEncoder module takes TransformerEncoderOptions in its constructor.  And within TransformerEncoderOptions,  it has a variable of TransformerEncoderLayer class since TransformerEncoder need TransformerEncoderLayer as input. If putting everything into one header file transformer.h. Then all the `Options` classes are within options/transformer.h, and all the `Module` classes are within modules/transformer.h, then options/transformer.h will need to include modules/transformer.h and vise versa (since TransformerEncoderOptions need TransformerEncoderLayer and TransformerEncoder need TransformerEncoderOptions). To avoid such cycle, we split the header files.

zou3519 replied:
Yeah, that's a good reason to have a separate header. Is there a reason why it is named "transformercoder.h" instead of "transformerencoder.h"?

glaringlee replied:
> Yeah, that's a good reason to have a separate header. Is there a reason why it is named "transformercoder.h" instead of "transformerencoder.h"?

Since there will be decoder code within this file as well. We have some external contributor working on decoder part

zou3519 commented on torch/csrc/api/include/torch/nn/modules/transformercoder.h:
```diff
@@ -0,0 +1,71 @@
+
+    Tensor forward(
+      const Tensor& src,
+      const Tensor& src_mask = {},
+      const Tensor& src_key_padding_mask = {});
```
Is the convention in the C++ API to use undefined tensors instead of `optional<Tensor>`? (Asking for my understanding)

glaringlee replied:
Yes. And actually the meaning here is that, this is default to undefined tensor not optional to undefined tensor.

zou3519 commented on torch/csrc/api/include/torch/nn/modules/transformercoder.h:
```diff
@@ -0,0 +1,71 @@
+
+    Tensor forward(
+      const Tensor& src,
+      const Tensor& src_mask = {},
```
The second argument is called "mask" in Python: https://github.com/pytorch/pytorch/blob/cbdaa20c88306b9a828edd6e766d3c987b19a628/torch/nn/modules/transformer.py#L167. Does it matter that this one is different?

I do prefer `src_mask` because it makes it clear what is getting masked...

glaringlee replied:
@zhangguanheng66 I make the name to src_mask for the same reason as @zou3519, agree?

zhangguanheng66 replied:
I think it makes sense to call `src_mask` and `tgt_mask` along with `src_key_padding_mask` and `tgt_key_padding_mask` (consistent with the python ones)

zou3519 commented on torch/csrc/api/include/torch/nn/modules/transformercoder.h:
```diff
@@ -0,0 +1,71 @@
+      const Tensor& src_mask = {},
+      const Tensor& src_key_padding_mask = {});
+
+    void reset() override;
+
+    void reset_parameters();
```
Sorry for the n00b question, but is there a guide for how to write C++ modules somewhere? I don't understand what these two are used for

glaringlee replied:
reset_parameter is not always called. This is a convenient function to reset all its parameters to the default status, some module has this function since they have private parameters other than parameters that registered to base Module class, or sometimes, user just want to reset the modules' parameters instead of re-create them. This is not always required though. We are thinking about putting this function as virtual in base class, but still under discussion since virtual will slowdown the code.

zou3519 replied:
makes sense, thanks for the clarification

zou3519 commented on torch/csrc/api/include/torch/nn/options/transformercoder.h:
````diff
@@ -0,0 +1,38 @@
+  /// auto options = TransformerEncoderOptions(encoderLayer, 6).norm(LayerNorm(LayerNormOptions({2})));
+  /// ```
+  struct TORCH_API TransformerEncoderOptions {
+    // This constructor will keep the a ref of passed in encoder_layer, so it keeps all the data in encoder_layer.
+    TransformerEncoderOptions(TransformerEncoderLayer encoder_layer, int64_t num_layers);
+    // This constructor will create a new TransformerEncoderLayer obj based on passed in encoder_layer_options.
+    TransformerEncoderOptions(const TransformerEncoderLayerOptions& encoder_layer_options, int64_t num_layers);
````
Are there any other Options that hold modules? (So I can calibrate my expectations against those)

glaringlee replied:
:), no, this is the first time, a module need another module as an input parameter, I had the same question as you, and looked around for a while, did not find anyone. So we are on our own.

zou3519 commented on torch/csrc/api/src/nn/modules/transformer.cpp:
```diff
@@ -223,5 +224,66 @@ Tensor TransformerDecoderLayerImpl::activation(const Tensor& input){
+  for (size_t i = 0; i < num_layers; ++i) {
+    layers->at<TransformerEncoderLayerImpl>(i).reset_parameters();
+  }
+  // a. No way to know whether module in AnyModule has api to reset_parameters, so replace instead
+  // b. Allow user to add/delete normalization module when reset parameters
```
When does `reset_parameters()` get called? is it a user API or something that is used in the framework?

glaringlee replied:
As I mentioned before, this is a convenient function to reset parameters within the module. It is not necessary to call it. And I believe, it won't be called in most of the case.

zou3519 replied:
Hmm, so the thing I am confused about here is:
- there's no reset_parameters() or _reset_parameters on TransformerEncoderLayer in Python. Some of our Python nn.Modules do have reset_parameters() that re-initializes parameters.

And one potential concern is, let's say you have an optimizer that takes in a list of parameters (disclaimer: I have no clue how optimizers in C++ work). Is the following possible?
- I create a TransformerEncoderLayer in C++
- I add pass the parameters of a TransformerEncoderLayer to an optimizer (in C++)
- then, I call reset_parameters() on TransformerEncoderLayer
- The optimizer holds onto references of the original parameters (and not the new ones!)

glaringlee replied:
This is not possible since the C++ module underlying is a shared pointer, and optimizer doesn't clone the module but just shallow copy the module, so this case should be ok.

But the reset_parameters() it self is tricky, I admit that, even in python impl, reset_parameters() has different usage in different modules and not always there and not always get called. There is a PR that trying to standarize it by making reset_parameter() virtual and moved that to base class, but this change doesn't make sense since half of the module doesn't have it and virtual will make extra cost. So currently, reset_parameters() is just an optional functions.

zou3519 commented on test/cpp/api/transformer.cpp:
```diff
@@ -436,7 +402,178 @@ TEST_F(TransformerTest, TransformerDecoderLayer_gelu) {
-    transformer_decoder_layer_test_helper_gelu(true);
+  transformer_decoder_layer_test_helper_gelu(true);
+}
+
```
Is there a reason why we're not flipping the parity for TransformerEncoder on https://github.com/pytorch/pytorch/blob/master/test/cpp_api_parity/parity-tracker.md ?

glaringlee replied:
Ah, yes, we will flip the parity. There will be a separate PR to enable all parities later. We are currently focusing on implementation and will resolve parity in later PRs.

zou3519 reviewed (commented):
(not a full review) asking some questions

zou3519 commented on torch/csrc/api/include/torch/nn/options/transformercoder.h:
````diff
@@ -0,0 +1,38 @@
+  /// auto options = TransformerEncoderOptions(encoderLayer, 6).norm(LayerNorm(LayerNormOptions({2})));
+  /// ```
+  struct TORCH_API TransformerEncoderOptions {
+    // This constructor will keep the a ref of passed in encoder_layer, so it keeps all the data in encoder_layer.
+    TransformerEncoderOptions(TransformerEncoderLayer encoder_layer, int64_t num_layers);
````
nit: "will keep the a ref of passed in" -> "will keep a ref of passed in"

Actually, it's not really that it is keeping a reference, it's more that it takes ownership of the thing passed-by-value, right?

glaringlee replied:
I will make it more clear, the reason I say it is still a ref because they share the same storage

glaringlee replied:
Changed to `shallow copy`

zou3519 commented on torch/csrc/api/src/nn/modules/transformer.cpp:
```diff
@@ -223,5 +224,66 @@ Tensor TransformerDecoderLayerImpl::activation(const Tensor& input){
+  reset();
+}
+
+void TransformerEncoderImpl::reset() {
+  layers = this->register_module("layers", ModuleList());
+  for (int64_t i = 0; i < options.num_layers(); ++i) {
+    layers->push_back(options.encoder_layer()->clone());
+  }
```
I feel a bit weird about each layer having a copy of the same parameters, but it looks like that is the same thing as what we do in Python

glaringlee replied:
@zou3519 This is what transformer does. It followed the the method in a paper here:
https://papers.nips.cc/paper/7181-attention-is-all-you-need.pdf
I am not sure the reason behind. Maybe @zhangguanheng66 can explain here.

zhangguanheng66 replied:
TransformerEncoder is composed of multiple layers and each layer is independent from the others. During training, the learnable parameters on each layers will be trained differently. For the python one, we pass a transformer encoder layer to the constructor of the transformer encoder and make deep copy. We just initiate the layers with same parameters but this is not necessary.

Commit:
Update on "C++ APIs TransformerEncoder"


Differential Revision: [D23182770](https://our.internmc.facebook.com/intern/diff/D23182770)

[ghstack-poisoned]

Commit:
Update on "C++ APIs TransformerEncoder"


Differential Revision: [D23182770](https://our.internmc.facebook.com/intern/diff/D23182770)

[ghstack-poisoned]

glaringlee commented:
@zhangguanheng66 
Do you agree on this change?
https://github.com/pytorch/pytorch/pull/43187#discussion_r476540331
Any comment on this?
https://github.com/pytorch/pytorch/pull/43187#discussion_r476674073

Thanks

zou3519 commented on torch/csrc/api/include/torch/nn/options/transformercoder.h:
```diff
@@ -0,0 +1,38 @@
+  struct TORCH_API TransformerEncoderOptions {
+    // This constructor will keep a shallow copy of encoder_layer, so it keeps all the data in encoder_layer.
+    TransformerEncoderOptions(TransformerEncoderLayer encoder_layer, int64_t num_layers);
+    // This constructor will create a new TransformerEncoderLayer obj based on passed in encoder_layer_options.
+    TransformerEncoderOptions(const TransformerEncoderLayerOptions& encoder_layer_options, int64_t num_layers);
```
this also keeps a shallow copy of the data in encoder_layer, right?

glaringlee replied:
@zou3519 
No, TransformerEncoder will use encoder_layer_options as input, allocating a new TransformerEncdoerLayer.

zou3519 replied:
Oh I see now, thanks for the clarification

Commit:
Update on "C++ APIs TransformerEncoder"


Differential Revision: [D23182770](https://our.internmc.facebook.com/intern/diff/D23182770)

[ghstack-poisoned]

Commit:
Update on "C++ APIs TransformerEncoder"


Differential Revision: [D23182770](https://our.internmc.facebook.com/intern/diff/D23182770)

[ghstack-poisoned]
